File: E2EAppFlutter/lib/screens/screens.dart
import 'package:flutter/material.dart';

import '../main.dart';
import '../tags.dart';
import '../widgets.dart';

class HomeScreen extends StatelessWidget {
  const HomeScreen({required this.onNavigate, super.key});

  final void Function(Screen) onNavigate;

  @override
  Widget build(BuildContext context) => ScreenColumn(children: [
        const TaggedText(Tags.homeMarker, 'E2E ホーム'),
        TaggedButton(Tags.navSelector, 'セレクタ',
            fillWidth: true, onTap: () => onNavigate(Screen.selector)),
        TaggedButton(Tags.navInput, 'テキスト入力',
            fillWidth: true, onTap: () => onNavigate(Screen.input)),
        TaggedButton(Tags.navGesture, 'ジェスチャ',
            fillWidth: true, onTap: () => onNavigate(Screen.gesture)),
        TaggedButton(Tags.navScroll, 'スクロール',
            fillWidth: true, onTap: () => onNavigate(Screen.scroll)),
        TaggedButton(Tags.navAsync, '非同期表示',
            fillWidth: true, onTap: () => onNavigate(Screen.async)),
        TaggedButton(Tags.navDialog, 'ダイアログ',
            fillWidth: true, onTap: () => onNavigate(Screen.dialog)),
        TaggedButton(Tags.navLifecycle, 'ライフサイクル',
            fillWidth: true, onTap: () => onNavigate(Screen.lifecycle)),
        TaggedButton(Tags.navHeal, '自己修復',
            fillWidth: true, onTap: () => onNavigate(Screen.heal)),
        TaggedButton(Tags.navDiagnostics, '診断',
            fillWidth: true, onTap: () => onNavigate(Screen.diagnostics)),
      ]);
}

class AboutScreen extends StatelessWidget {
  const AboutScreen({super.key});

  @override
  Widget build(BuildContext context) => const ScreenColumn(children: [
        TaggedText(Tags.txtAboutMarker, 'E2E について'),
        TaggedText(Tags.txtAboutApp, 'app=${AppInfo.appId}'),
        TaggedText(Tags.txtAboutVersion, 'version=${AppInfo.version}'),
      ]);
}

class SelectorScreen extends StatefulWidget {
  const SelectorScreen({super.key});

  @override
  State<SelectorScreen> createState() => _SelectorScreenState();
}

class _SelectorScreenState extends State<SelectorScreen> {
  String _result = '-';

  void _set(String v) => setState(() => _result = v);

  @override
  Widget build(BuildContext context) => ScreenColumn(children: [
        TaggedText(Tags.selectorResult, 'result=$_result'),
        // 「許可」⊂「通知を許可」の部分一致衝突は契約で意図的に作られた検証材料。
        TaggedButton(Tags.btnAllow, '許可', onTap: () => _set('allow')),
        TaggedButton(Tags.btnAllowNotification, '通知を許可',
            onTap: () => _set('allow_notification')),
        // 同一ラベル「項目」の3連。ラベル指定では曖昧・#id か .Type[n] でのみ引ける。
        TaggedButton(Tags.btnItem1, '項目', onTap: () => _set('item1')),
        TaggedButton(Tags.btnItem2, '項目', onTap: () => _set('item2')),
        TaggedButton(Tags.btnItem3, '項目', onTap: () => _set('item3')),
        const TaggedText(Tags.txtSharedLabel, '共通ラベル'),
        TaggedButton(Tags.btnSharedLabel, '共通ラベル', onTap: () => _set('shared')),
        TaggedButton(Tags.btnAliasNew, '別名ボタン', onTap: () => _set('alias')),
        TaggedButton(Tags.btnSelectorReset, '結果クリア', onTap: () => _set('-')),
        // 700dp: 初期表示では絶対に画面内に入らない高さ(scrollTo / requireVisible の検証材料)。
        const SizedBox(height: 700),
        const TaggedText(Tags.txtOffscreen, '画面外テキスト'),
      ]);
}

// レイアウトはソフトキーボードに支配される(Compose 版と同じ制約)。スクロールさせず、
// シナリオが触る要素(echo 4本 + 単一行/パスワード欄 + 送信/クリア)を画面上部に固める。
class InputScreen extends StatefulWidget {
  const InputScreen({super.key});

  @override
  State<InputScreen> createState() => _InputScreenState();
}

class _InputScreenState extends State<InputScreen> {
  final _single = TextEditingController();
  final _password = TextEditingController();
  final _multiline = TextEditingController();
  String _submitted = '-';

  @override
  void dispose() {
    _single.dispose();
    _password.dispose();
    _multiline.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) => ScreenColumn(scrollable: false, children: [
        TaggedText(Tags.echoSingle, 'single=${_single.text}'),
        TaggedText(Tags.echoPassword, 'password=${_password.text}'),
        TaggedText(Tags.echoLength, 'len=${_single.text.length}'),
        TaggedText(Tags.txtInputSubmitted, 'submitted=$_submitted'),
        tagged(
          Tags.fieldSingle,
          TextField(
            controller: _single,
            onChanged: (_) => setState(() {}),
            decoration: const InputDecoration(hintText: '単一行'),
          ),
        ),
        tagged(
          Tags.fieldPassword,
          TextField(
            controller: _password,
            obscureText: true,
            onChanged: (_) => setState(() {}),
            decoration: const InputDecoration(hintText: 'パスワード'),
          ),
        ),
        Row(children: [
          Expanded(
            child: TaggedButton(Tags.btnInputSubmit, '送信',
                fillWidth: true, onTap: () => setState(() => _submitted = _single.text)),
          ),
          const SizedBox(width: 8),
          Expanded(
            child: TaggedButton(Tags.btnInputClear, '入力クリア', fillWidth: true, onTap: () {
              _single.clear();
              _password.clear();
              _multiline.clear();
              setState(() => _submitted = '-');
            }),
          ),
        ]),
        TaggedText(Tags.echoMultiline, 'multiline=${_multiline.text.replaceAll('\n', ' ')}'),
        tagged(
          Tags.fieldMultiline,
          TextField(
            controller: _multiline,
            maxLines: 3,
            onChanged: (_) => setState(() {}),
            decoration: const InputDecoration(hintText: '複数行'),
          ),
        ),
      ]);
}

class LifecycleScreen extends StatefulWidget {
  const LifecycleScreen({super.key});

  @override
  State<LifecycleScreen> createState() => _LifecycleScreenState();
}

class _LifecycleScreenState extends State<LifecycleScreen> {
  @override
  Widget build(BuildContext context) => ScreenColumn(children: [
        TaggedText(Tags.txtLaunchCount, 'launch=${LaunchCounter.value}'),
        TaggedText(Tags.txtSessionCount, 'session=${SessionCounter.value}'),
        TaggedButton(Tags.btnSessionInc, 'セッション+1',
            onTap: () => setState(() => SessionCounter.value += 1)),
        TaggedButton(Tags.btnResetPersisted, '永続カウンタをリセット', onTap: () async {
          await LaunchCounter.reset();
          if (mounted) setState(() {});
        }),
        TaggedText(Tags.txtPlatform,
            'platform=${Theme.of(context).platform == TargetPlatform.iOS ? 'iOS' : 'Android'}'),
      ]);
}

class HealScreen extends StatefulWidget {
  const HealScreen({super.key});

  @override
  State<HealScreen> createState() => _HealScreenState();
}

class _HealScreenState extends State<HealScreen> {
  late bool _schemaV1 = prefs.getBool('heal_schema_v1') ?? true;
  String _tapped = '-';

  @override
  Widget build(BuildContext context) => ScreenColumn(children: [
        Row(mainAxisSize: MainAxisSize.min, children: [
          const Text('旧ID(v1)を使う'),
          tagged(
            Tags.swHealSchema,
            Switch(
              value: _schemaV1,
              onChanged: (v) async {
                await prefs.setBool('heal_schema_v1', v);
                setState(() => _schemaV1 = v);
              },
            ),
          ),
        ]),
        TaggedText(Tags.txtHealSchema, 'schema=${_schemaV1 ? 'v1' : 'v2'}'),
        // ラベル固定・id のみ切替がヒール検証の核。
        TaggedButton(_schemaV1 ? Tags.btnHealV1 : Tags.btnHealV2, '修復対象',
            onTap: () => setState(() => _tapped = _schemaV1 ? 'v1' : 'v2')),
        TaggedText(Tags.txtHealResult, 'tapped=$_tapped'),
        TaggedButton(Tags.btnHealReset, '修復結果クリア',
            onTap: () => setState(() => _tapped = '-')),
      ]);
}

File: E2EAppAndroid/app/src/main/kotlin/com/ftester/e2e/android/Screens.kt
package com.ftester.e2e.android

import android.app.Activity
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView

private fun Activity.inflate(layout: Int, parent: ViewGroup): View =
    layoutInflater.inflate(layout, parent, false)

fun buildHomeScreen(activity: Activity, parent: ViewGroup, onNavigate: (Screen) -> Unit): View {
    val v = activity.inflate(R.layout.screen_home, parent)
    v.findViewById<Button>(R.id.nav_selector).setOnClickListener { onNavigate(Screen.SELECTOR) }
    v.findViewById<Button>(R.id.nav_input).setOnClickListener { onNavigate(Screen.INPUT) }
    v.findViewById<Button>(R.id.nav_gesture).setOnClickListener { onNavigate(Screen.GESTURE) }
    v.findViewById<Button>(R.id.nav_scroll).setOnClickListener { onNavigate(Screen.SCROLL) }
    v.findViewById<Button>(R.id.nav_async).setOnClickListener { onNavigate(Screen.ASYNC) }
    v.findViewById<Button>(R.id.nav_dialog).setOnClickListener { onNavigate(Screen.DIALOG) }
    v.findViewById<Button>(R.id.nav_lifecycle).setOnClickListener { onNavigate(Screen.LIFECYCLE) }
    v.findViewById<Button>(R.id.nav_heal).setOnClickListener { onNavigate(Screen.HEAL) }
    v.findViewById<Button>(R.id.nav_diagnostics).setOnClickListener { onNavigate(Screen.DIAGNOSTICS) }
    return v
}

fun buildSelectorScreen(activity: Activity, parent: ViewGroup): View {
    val v = activity.inflate(R.layout.screen_selector, parent)
    val result = v.findViewById<TextView>(R.id.txt_selector_result)
    fun set(value: String) { result.text = "result=$value" }

    v.findViewById<Button>(R.id.btn_allow).setOnClickListener { set("allow") }
    v.findViewById<Button>(R.id.btn_allow_notification).setOnClickListener { set("allow_notification") }
    v.findViewById<Button>(R.id.btn_item_1).setOnClickListener { set("item1") }
    v.findViewById<Button>(R.id.btn_item_2).setOnClickListener { set("item2") }
    v.findViewById<Button>(R.id.btn_item_3).setOnClickListener { set("item3") }
    v.findViewById<Button>(R.id.btn_shared_label).setOnClickListener { set("shared") }
    v.findViewById<Button>(R.id.btn_alias_new).setOnClickListener { set("alias") }
    v.findViewById<Button>(R.id.btn_selector_reset).setOnClickListener { set("-") }
    return v
}

fun buildInputScreen(activity: Activity, parent: ViewGroup): View {
    val v = activity.inflate(R.layout.screen_input, parent)
    val single = v.findViewById<EditText>(R.id.field_single)
    val password = v.findViewById<EditText>(R.id.field_password)
    val multiline = v.findViewById<EditText>(R.id.field_multiline)
    val echoSingle = v.findViewById<TextView>(R.id.txt_echo_single)
    val echoPassword = v.findViewById<TextView>(R.id.txt_echo_password)
    val echoMultiline = v.findViewById<TextView>(R.id.txt_echo_multiline)
    val echoLength = v.findViewById<TextView>(R.id.txt_echo_length)
    val submitted = v.findViewById<TextView>(R.id.txt_input_submitted)

    single.onTextChanged {
        echoSingle.text = "single=$it"
        echoLength.text = "len=${it.length}"
    }
    // パスワードは平文で echo する(検証用。実アプリでは絶対にやらない)。
    password.onTextChanged { echoPassword.text = "password=$it" }
    multiline.onTextChanged { echoMultiline.text = "multiline=${it.replace("\n", " ")}" }

    v.findViewById<Button>(R.id.btn_input_submit).setOnClickListener {
        submitted.text = "submitted=${single.text}"
    }
    v.findViewById<Button>(R.id.btn_input_clear).setOnClickListener {
        single.setText("")
        password.setText("")
        multiline.setText("")
        submitted.text = "submitted=-"
    }
    return v
}

private fun EditText.onTextChanged(block: (String) -> Unit) {
    addTextChangedListener(object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
        override fun afterTextChanged(s: Editable?) { block(s?.toString() ?: "") }
    })
}

fun buildLifecycleScreen(activity: Activity, parent: ViewGroup): View {
    val v = activity.inflate(R.layout.screen_lifecycle, parent)
    val launch = v.findViewById<TextView>(R.id.txt_launch_count)
    val session = v.findViewById<TextView>(R.id.txt_session_count)
    fun render() {
        launch.text = "launch=${LaunchCounter.value}"
        session.text = "session=${SessionCounter.value}"
    }
    v.findViewById<Button>(R.id.btn_session_inc).setOnClickListener {
        SessionCounter.value += 1
        render()
    }
    v.findViewById<Button>(R.id.btn_reset_persisted).setOnClickListener {
        LaunchCounter.reset()
        render()
    }
    render()
    return v
}

fun buildHealScreen(activity: Activity, parent: ViewGroup): View {
    val v = activity.inflate(R.layout.screen_heal, parent)
    val sw = v.findViewById<androidx.appcompat.widget.SwitchCompat>(R.id.sw_heal_schema)
    val schema = v.findViewById<TextView>(R.id.txt_heal_schema)
    val v1 = v.findViewById<Button>(R.id.btn_heal_v1)
    val v2 = v.findViewById<Button>(R.id.btn_heal_v2)
    val result = v.findViewById<TextView>(R.id.txt_heal_result)

    // ラベルは不変(「修復対象」)で id だけが入れ替わるのがヒール検証の核。
    fun render(schemaV1: Boolean) {
        schema.text = "schema=${if (schemaV1) "v1" else "v2"}"
        v1.visibility = if (schemaV1) View.VISIBLE else View.GONE
        v2.visibility = if (schemaV1) View.GONE else View.VISIBLE
    }

    sw.isChecked = Prefs.getBool("heal_schema_v1", true)
    render(sw.isChecked)
    sw.setOnCheckedChangeListener { _, checked ->
        Prefs.setBool("heal_schema_v1", checked)
        render(checked)
    }
    v1.setOnClickListener { result.text = "tapped=v1" }
    v2.setOnClickListener { result.text = "tapped=v2" }
    v.findViewById<Button>(R.id.btn_heal_reset).setOnClickListener { result.text = "tapped=-" }
    return v
}
